### ISSUE TITLE
Process for minor or bug fix changes?

### ISSUE THREAD

**Issue by michaelmhoffman** [ga4gh/product-process#2]
What are the kinds of changes to standards that do not require going through the full process? For example:

- minor changes that don't result in backwards compatibility issues
- clarifying mostly obvious minor errors that might have implementation ramifications if misinterpreted
- fixing typographical errors that won't have implementation ramifications
- changes to annexes
- changes to test suites

**Comment by susanfairley** [comment]
Thanks Michael - this isn't specified in the current draft and should be.

My thinking is bug fixes and minor error corrections should be quick - especially where not acting may cause problems for implementors and adoption is currently limited.

Minor changes (no backward compatibility issues) need some process and I'm interested in using a light touch version of this one, so that there is both some form of review and a documented process. In that instance, I would expect the review to focus on what was new and to reuse, wherever possible, existing documentation.

To elaborate, I'm not convinced that changing an existing standard (even with backward compatibility) should be easier than setting up a new one. I also think existing users should be being consulted on planned changes. In addition, other things - like testing interoperability, review of the specification, etc are, I think, still applicable to minor changes. If the changes are minor, however, I'd be very happy to see reuse of existing documentation and other options for minimising workload being used.

Arguably, this also relates to stability and maturity of the products. There has been recent discussion around using 'trial use' for products that are still in the early stages. This has been discussed in GKS (@larrybabb, @ahwagner and @andrewyatz).

Thanks again for raising this issue @michaelmhoffman. I'd love to get more comments on this.

**Comment by jkbonfield** [comment]
As a concrete example, adding a new SAM tag type is an addition to a specification.  It is however an additional chunk of documentation, which needs at least some approval.  Until now though it's been considered that approval amongst the other spec maintainers was sufficient, and this is what to us our "light touch" entails.  That doesn't mean they're always fast - barcoding and methylation tags took years mainly due to lack of community cohesion (no matter how necessary it is).  But sometimes they can be quite simple and obvious too.  We did however add the methylation tags as an unofficial (private name space) tag for a while as a trial period, and that seemed to unblock the catch-22 of no support because not in spec, and not in spec because of no support.

Another example was the discussion on adding BGI vs MGI vs MGISEQ etc as a platform type in the read-group SAM header lines.  It's a formal part of the spec, but clearly does not impact on any previous files.

At the most minor level, reformatting of tables, maybe adding a clarification footnote or changing a phrase which later turns out to be a little bit ambiguous, are things we do regularly and pretty much it's just verbal agreement on a conference call and then it's done and dusted.  Any formal process there will just make people less inclined to fix these issues.

**Comment by andrewyatz** [comment]
Yes GKS has been discussing ideas around stability of standards and their adoption in the wider community. The less well adopted/mature a product is the more you can expect it to change or be retired. Trial use would be one of those options where something hasn't yet matured enough

**Comment by andrewyatz** [comment]
I feel that @jkbonfield is showing a good example of something which would benefit from an in-depth review of the study group/further approval of the standard before being marked as "live" or "mature". I can't help but look back at the examples @michaelmhoffman as the other end of this where there's a number of elements there that shouldn't require any additional work/approval e.g. typo corrections.

Perhaps the most worrying part of the diagram is the study group review by SC. If a PRC is empowered by SC to monitor a standard's study group perhaps they are best placed to recommend presentation and approval at SC? This perhaps would allow for a balance to be struck between the two update examples presented here

**Comment by jkbonfield** [comment]
Thinking of study groups, I also wonder just who all these people are going to be?  We simply don't have the body count to act as spec maintainers right now (we've been lacking an active 2nd CRAM maintainer for years, and only very recently managed to plug gaps in VCF).  To expect a "study group" to be found in addition to this which isn't exactly the same people as we already have is idealistic maybe.

I'd *love* for it to work, but if so GA4GH will need to much more proactive in cajoling people to step forward and help out.  Otherwise we just run the risk of things stalling indefinitely.  I don't have any magic solutions for this sadly.

**Comment by ahwagner** [comment]
I had similar concerns to @jkbonfield. For example, in VICC we are currently working with a large clinical community (representatives from CAP, ACMG, CGC, ClinGen, and VICC) to define the minimal data elements for Gene Fusions–an effort that has taken years to build consensus guidelines, which I expect will ultimately inform the development of a corresponding data exchange model in VRS. The development of that model will then have benefited from a clinical community to drive the data requirements, and a data modeling community (the GKS VR team) to develop the nuts and bolts for the standard. It is unclear how a study group effectively complements this expertise.

Is there any way that existing efforts to build community consensus (such as the above cross-consortia study) can be used as a proposal in lieu of a new _ad hoc_ study group? Perhaps reviewed and approved by Work Stream leads prior to further development work? And leave study group formation as an option to be exercised by those leads if they feel the groundwork for the proposal is insufficient / additional expertise is required?

**Comment by andrewyatz** [comment]
@susanfairley I think so. It was more me thinking in terms of the point/responsibilities of the PRC. It feels like, to me, PRC is formed and empowered by SC to ensure study groups remain on-track and their discussions are relevant. That process of oversight seems somewhat immaterial to the actual outputs. This would also be more inline with the later SC "vote" which from memory you said would approve that process has been followed. 

Overall we need to get the balance right here about what SC has a say about and how it has that say

**Comment by jkbonfield** [comment]
What's the minimum size for a study group?  Eg could it just be 2 people?

**Comment by susanfairley** [comment]
@jkbonfield - it could be just two people as currently described (no minimum is specified) - but they'd need to be two people who built a convincing case that their work would be broadly useful.

**Comment by aniewielska** [comment]
A similar question asked in #10. I believe the current process used in the Cloud Workstream for minor changes was not to form PRC or involve SC, so this would be a change for us.

**Comment by jmarshall** [comment]
Here are several more concrete examples of different levels of change, from trivial to major, in a SAM/BAM/etc context:

1. Typo fixes or other editorial trivia such as formatting (to text or diagrams), with no effect on the meaning of the text.
2. Clarification of existing unclear text, where everyone [^1] agrees on the intended meaning.
3. Clarification of behaviour not (sufficiently) explicitly covered in the text, where everyone agrees on the intended behaviour
4. Clarification of behaviour not (sufficiently) explicitly covered in the text, where there is disagreement as to the intended behaviour
5. Adding a new simple SAM tag, for example a description tag. (Note that SAM is a format that was explicitly designed to be compatibly extensible — which is a concept that may warrant being reflected in some flowchart choices.)
6. Defining a whole swathe of tags that break new ground, defining a new representation of some larger chunk of data; e.g., barcode tags or methylation, as mentioned in previous comments.
7. Making a new incompatible variant of a format — e.g. revising 32-bit fields into 64 bits and coming up with BAMv2.

Clearly (1) and arguably clearly (2) and (3) should not require additional process, while anything beyond (7) should. Where the dividing line lies in between and with what graduations of grey is an open question IMHO.

[^1]: “Everyone” being e.g. everyone with sufficient knowledge of and familiarity with the subject matter to have a well-founded opinion.

**Comment by michaelmhoffman** [comment]
I agree with @jmarshall that 1 definitely should not require additional process and 7 definitely should. For some stuff in the middle, "light process" may be helpful, as @susanfairley suggested. It is definitely worth considering that heavy process is going to make it less likely that 1–4 are fixed, as @jkbonfield. That would be a bad outcome.

**Comment by susanfairley** [comment]
Thanks to everyone for the feedback so far. I'd like to use this issue as the venue to discuss how we make updates to existing standards and would like to pull in #10, which had contributions from @aniewielska and @briandoconnor

**Comment by susanfairley** [comment]
Again, thanks to everyone who has commented here. A few thoughts to add on this:

### Study group -> Study phase
I wonder if it is better to relabel the Study Group as a Study Phase to help clarify that this need not be a different group but that there is an expectation that people first establish that there is a problem worth solving, and that that problem is agreed and defined, before progressing.

### Categories of change
Thinking about change considerations and level of process needed, I'm finding myself thinking about three categories.

1. **Urgent fix** - these could include bugs and time sensitive changes (I'm thinking of a case where an error was spotted in recently approved documentation and where making the change promptly limited the impact). These would be changes 'in the spirit of' the previously approved standard but could include things that are technically breaking changes. This would hopefully be used infrequently.
2. **Continued development** - again 'in the spirit of' the previously approved standard and excluding breaking changes. This could include ongoing maintenance, clarification and comparatively minor new additions. Looking at the SAM example above, I'd imagine 1-5 easily falling in this category, 6 debatable and 7, as a breaking change, would be outside it.
3. **Scope change or breaking change** - a better name would be nice but I'd see this as changes that break backward compatibility, add significant new features that were not present before (so maybe 6 from the above SAM example) and/or take the standard in a new direction

### 'in the spirit of'
To elaborate on what I mean with these words, I'd see it as something that someone well informed about the standard would see as fitting naturally with the last version of the standard to go through Steering Committee approval. To return to the SAM example above, I'd be happy about 5 being included here (SAM is intended to be extensible and I'm interpreting 'simple' as non-contentious).

### Breaking changes

1. I think breaking backward compatibility for an approved standard (possible exceptions for draft and trial standards) should be relatively hard to do.
2. Before breaking changes are made, I think there should be consultation with those who have already adopted the standard (possible exception for urgent fixes).

### Draft and trial
These weren't part of our process before but GKS have been looking at them. My thinking is that, for changes to a standard labelled as draft or trial, the continued development mode would be reasonable, as breaking changes should be anticipated.

What is more of a question for me is how are these standards approved in the first place? I find it difficult to argue against the Study phase to determine if a standard is of use, so I'm not sure that I regard that as optional.

### Decision making
The question of which category a change comes under would, I think, be best taken by the subgroup and PRC. My reasons for not leaving this solely to the subgroup are that some of our groups are small and not all have well defined decision making processes. I think this leaves them vulnerable, in some cases, to attempts to push work through. With that in mind, I prefer a model where there is some 'light' external oversight from the PRC. In all cases, I think we need to move to decision making processes being defined for all groups.

### Broad agreement
Above, in defining something, I found myself falling back on saying if it's not 'contentious'. With a possible exception for urgent fixes, if we don't see broad agreement, then I don't think we should be pushing something through - it may need more scrutiny, time or work.

###Process ideas
To take this further, some (rough) ideas on what process for the three cases suggested above might look like.

**Urgent fix**

1. An issue seeming to require an urgent fix comes to light. The PRC is given a heads up and the subgroup starts work on a fix (working publicly unless there is a security concern), assuming that the PRC agrees the urgent fix mechanism is appropriate for the case.
2. Group agrees a solution (using their decision making process).
3. Solution is posted publicly and passed to PRC for a short (seven day?) review period. PRC and subgroup leads can override this if it is judged to be clearcut and there is concern about the delay.
4. If no significant objections have been raised (as judged by the PRC and subgroup leads) the change goes through as a patch.

**Continued development**

1. An issue seeming to meet the continued development criteria is known. The subgroup, using their decision making mechanism, decide to work on it or not.
2. Work commences publicly and the PRC are informed that this work is proceeding on this track, with a means of objecting to the choice of track.
3. The group agrees a solution, ready for review, using their decision making mechanism.
4. The proposed revision is posted for public comment and review by the PRC (possibly one month?)
5. Once concerns have been addressed to the joint satisfaction of the subgroup and PRC, the revision goes through as a minor version.

**Scope change or breaking change**

1. The subgroup discusses the topic/feature/problem of interest and reaches a decision on proceeding or not, using their decision making mechanism.
2. The subgroup develops a problem statement (analogous to the study outputs in the main process) and consultation of existing adopters is mandatory (at a minimum driver projects using the standard but some discretion for the PRC to judge if adequate). The problem statement should make clear the need for the new work, who will use it and also be clear about what is not in scope.
3. Development work continues, similar to the proposed process.

This last version is, essentially the same as was proposed. It would encourage reuse of documentation to minimise work but would require evidence that the new work was needed and consultation with adopters. There would also be the Steering Committee review, if we went with this.

On reading this through, I realise I've arrived at a major, minor and patch scheme - so we could use those terms but I'll leave my words above as I hope they make my meaning clearer.

Interested in thoughts and comments. Thanks.

**Comment by susanfairley** [comment]
I'm hoping to draft some changes to reflect the above shortly. If you have an opinion on the above, please do add it.

**Comment by andrewyatz** [comment]
Not to take this in a different direction but the overall aims of this process appear to be:

1. Ensure that categories 1 and 2 of change (urgent and continued development) are not subject to an onerous process
2. Category 3 changes (scope or breaking) results in more appropriate oversight
3. Making sure that category 1 & 2 changes aren't actually mis-classified category 3 changes

I worry that some of the plan might still bring too much gating. I suggest we put more trust our community of product developers to know when they have category 3 changes to flag. I do like the idea that the PRC continues to be involved beyond the life of the first version and can continue to provide input into changes but if they fail to provide within sensible timelines the lack of feedback doesn't mean prevent to move forward.

Also the use of semantic versioning language is a good option IMO. I will say though this versioning scheme is really a TL;DR; of the release notes i.e. provider has decided the impact of a new version is scoped to a particular level. Note that doesn't mean a downstream toolkit couldn't break

**Comment by susanfairley** [comment]
Thanks for this @andrewyatz. I'm in general agreement.

One difference of opinion, however, on trust. I have huge respect for GA4GH contributors. In any context, however, I'm not persuaded that people marking their own homework is a good approach. For me, it's too close to "the good chap" theory - where we rely on people to do the right thing. Excellent though GA4GH contributors are, reality is that people cannot always be relied upon to do the right thing. As the standards are adopted more widely, these decisions have the potential to have significant consequences (in all sorts of ways - data sharing, privacy, ethics, commercial and financial). As such, as a professional organisation, a minimum level of oversight is required, in my opinion.

I also worry that leaving decisions with small working groups, without oversight, potentially places considerable pressure on group members to speak up in what may be challenging circumstances. The power balance and working dynamics in group settings can be challenging. While we have some excellent examples of wonderful group operation in GA4GH, I don't think we can assume that that will always be in place.

For PRC feedback, I'm open to specifying a timeframe and looking at wording to prevent things stalling.

**Comment by andrewyatz** [comment]
Each of these standards though works within a work stream, has organisational oversight from the work stream managers and two work stream leads. I hope that together gets us away from the "good chap" theory. 

Anyway with the PRC situation, can the gate be if there's not been a PRC response by timeframe X can we assume it's consent?

**Comment by susanfairley** [comment]
Some work streams have very clear documented policies and organisation but that is not the case across the board - there is a high level of variability, so it's not clear to me that it is guaranteed that that oversight is in place for all work streams.

I'd hesitate to say that no response equals consent but I would argue strongly that Secretariat should be working very actively to establish why there hasn't been a response. If it's not down to the content of the change but, say, due to a member of the PRC being unavailable, I'd certainly be open to letting two out of three members give something approval, etc, etc. I think this could be one where there could be many factors that could come into play and Secretariat should be working hard to make sure things don't get stalled.

**Comment by andrewyatz** [comment]
The stalling is my main concern and very likely IMO

**Comment by susanfairley** [comment]
I can see the concern and, to my mind, it's very reasonable. Will give further thought to ways to prevent such a step becoming a block.

**Comment by susanfairley** [comment]
Have not yet addressed everything discussed here but have now added flow diagram and text aiming to cover the urgent fix and continuing development cases. Will aim to review, noting that the study group -> study phase change has not yet been made. See commits immediately above for diagram and text alterations.
